crates/ethrpc/src/current_block/mod.rs

```diff
@@ -130,6 +128,41 @@ pub async fn current_block_stream(
     Ok(receiver)
 }
 
+/// Returns a stream that is synchronized to the passed in stream by only yields
+/// every nth update of the original stream.
+pub fn throttle(blocks: CurrentBlockStream, updates_to_skip: NonZeroU64) -> CurrentBlockStream {
+    let first_block = *blocks.borrow();
+    let (sender, receiver) = watch::channel(first_block);
+
+    let update_future = async move {
+        let mut skipped_updates = 0;
+        let mut block_stream = into_stream(blocks);
+
+        // Stream yields initial block immediately so we skip that on purpose.
+        let _ = block_stream.next().await;
```

Comment:
I don't think this is correct. Some logic requires that they get the block which is sure to be mined before anything after the first call to next happens (e.g. solution submission logic makes sure it gets the block number before it send out the settlement). If in that case we first wait for some updates we are breaking that assumption.

I think a throttled block stream should have the same behaviour (yield current block first) and then only update every nth block.

Reply:
If I understand your concern correctly exactly that is asserted [here](https://github.com/cowprotocol/services/pull/2678/files#diff-dde10be91d945474ef28b38bd7c06a2865cb4d54bd901511a9841d513d7d1621R379-R382).

The reason we have to do it like this is that we build a new stream with the current block of the original stream. This is the block that gets yielded right away. However, when we call `into_stream()` and await the next block we also get the current block because of the behavior you mentioned. So if we don't skip that block we would yield the original block twice.